software/displayerang.livecodescript

```diff
@@ -0,0 +1,1402 @@
+   end repeat
+   -- ! deal with elements that are visible but weren't discovered by middleOut
+   repeat for each word tKind in "nodes links"
+      repeat for each line tRec in the keys of pData[tKind]
+         if tRec is not among the items of pData["visited"][tKind] then
+            put tRec into item (the number of items of pData["unvisited"][tKind])+1 of pData["unvisited"][tKind]
+         end if
+      end repeat
+   end repeat
+   put pData into pDraw
+end middleOut
+
+on assignLayers @pAll,pSeed,pTheLayer -- ! the recursive part of middleOut
+   put "on assignLayers" into xX["write"];    devLog xX
+   -- pAll = pData (just a copy of sGraphData at the moment); this adds display info to it
+   -- pSeed = the node to find neighbors of
+   -- pTheLayer = the layer pSeed was assigned before this function was called
+   --- list all next & prev links
+   --- format as links
+   --- don't add if already on list of visited links
+   --- add node to layer; prev -1; next +1
+   --- call this function again on this node
+   --- terminal case is when this node has no unvisited links
+   if pSeed is not among the items of pAll["visited"]["nodes"] then
+      put pSeed & "," after pAll["visited"]["nodes"]
+      put pSeed & "," after pAll["layers"][pTheLayer]
+      put pTheLayer into pAll["nodes"][pSeed]["layer"]
+      repeat for each word tDir in "next prev"
+         --         repeat for each item tLink in pAll["nodes"][pSeed][tDir]
+         repeat for each line tLink in pAll["nodes"][pSeed][tDir]
+            if tLink is not among the items of pAll["visited"]["links"] then
+               put tLink & "," after pAll["visited"]["links"]
+               filter pAll["visited"]["links"] without empty
+               put item 1 of pAll["links"][tLink][tDir] into tNode
+               switch tDir
+                  case "next"
+                     assignLayers pAll,tNode,(pTheLayer +1)
+                     break
+                  case "prev"
+                     assignLayers pAll,tNode,(pTheLayer - 1)
+                     break
+               end switch
+            end if
+         end repeat
+      end repeat 
+   end if
+end assignLayers
+
+on displayManager pInfo,@pGUI,pSizes,pNow,@pHist -- ! creates/changes/deletes controls
+   put "on displayManager" into xX["write"];    devLog xX
+   -- sGraphDrawing,sGUIstuff,tSizes,sNowFocus,sMouseHistory
+   --pSizes[width/height/layer/node]
+   lock messages
+   -- delete old node graphics; brute force but not much to draw at the moment   
+   if pGUI is not empty then
+      repeat for each word tKind in "nodes links"
+         repeat for each key tNum in pGUI[tKind]
+            put pGUI[tKind][tNum]["graphic"] into tID
+            if exists(tID) then delete tID
+         end repeat
+      end repeat
+   end if
+   put pSizes["width"] into tWinWidth
+   put pSizes["height"] into tWinHeight
+   put tWinWidth/2 into tCenterH
+   lock screen
```

Comment:
@matthewmaier You may want to move `lock screen` to where you `lock messages` to be above the object deletions